### PR TITLE
[Merged by Bors] - feat(Data.Set.Ncard): enat set cardinality

### PR DESCRIPTION
**Pull request by apnelson1**
This PR is a second attempt at defining the cardinality of a set as extended natural number, with a function `encard`. The implementation involves a refactor, where the existing`ncard` function is redefined in terms of `encard`. This shortens a lot of proofs, reduces reliance on the `Finset` API, and allows for a potential future refactor where `ncard` is removed if it is decided to be redundant.  

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Ncard.lean [hunk 1]
```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
...
+
+variable {s : Set α} {t : Set β} {f : α → β}
+
+theorem encard_image_of_injOn (h : InjOn f s) : (f '' s).encard = s.encard := by
+  rw [encard, PartENat.card_image_of_injOn h, encard]
+
+theorem encard_congr (e : s ≃ t) : s.encard = t.encard := by
+  rw [←encard_univ_coe, ←encard_univ_coe t, encard_univ, encard_univ, PartENat.card_congr e]
+
+theorem encard_image_of_injective (hf : f.Injective) (s : Set α) : (f '' s).encard = s.encard :=
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:370]:
```suggestion
theorem _root_.Function.Injective.encard_image (hf : f.Injective) (s : Set α) : (f '' s).encard = s.encard :=
```

File: Mathlib/Data/Set/Ncard.lean [hunk 2]
```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
...
+  encard_image_of_injOn (hf.injOn s)
+
+theorem encard_le_of_embedding (e : s ↪ t) : s.encard ≤ t.encard := by
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:373]:
```suggestion
theorem _root_.Function.Embedding.encard_le (e : s ↪ t) : s.encard ≤ t.encard := by
```

File: Mathlib/Data/Set/Ncard.lean [hunk 3]
```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
...
+
+theorem Finite.injOn_of_encard_image_eq (hs : s.Finite) (h : (f '' s).encard = s.encard) :
+    InjOn f s := by
+  obtain (h' | hne) := isEmpty_or_nonempty α
+  · rw [s.eq_empty_of_isEmpty]; simp
+  rw [←encard_image_of_injOn (invFunOn_injOn_image f s)] at h
+  rw [injOn_iff_invFunOn_image_image_eq_self]
+  exact hs.eq_of_subset_of_encard_le (invFunOn_image_image_subset f s) h.symm.le
+
+theorem encard_preimage_of_injective_subset_range (hf : f.Injective) (ht : t ⊆ range f) :
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:393]:
```suggestion
theorem _root_.Function.Injective.encard_preimage_of_subset_range (hf : f.Injective) (ht : t ⊆ range f) :
```

File: Mathlib/Data/Set/Ncard.lean [hunk 4]
```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
...
+    (f ⁻¹' t).encard = t.encard := by
+  rw [←encard_image_of_injective hf, image_preimage_eq_inter_range,
+    inter_eq_self_of_subset_left ht]
+
+theorem encard_le_encard_of_injOn (hf : ∀ a ∈ s, f a ∈ t) (f_inj : InjOn f s) :
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:398]: `hf : MapsTo f s t`, and you probably want to rename this to take advantage of dot notation.

File: Mathlib/Data/Set/Ncard.lean [hunk 5]
```diff
@@ -56,47 +447,69 @@ syntax "toFinite_tac" : tactic
...
+  statements -/
+macro_rules
+  | `(tactic| to_encard_tac) => `(tactic|
+      simp only [←Nat.cast_le (α := ℕ∞), ←Nat.cast_inj (R := ℕ∞), Nat.cast_add, Nat.cast_one])
+
 
 /-- The cardinality of `s : Set α` . Has the junk value `0` if `s` is infinite -/
 noncomputable def ncard (s : Set α) :=
-  Nat.card s
-#align set.ncard Set.ncard
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:64]: Here and everywhere else: I think our plan is to leave the `#align`s around for a while (even if they are no longer perfect), so please put them back.

File: Mathlib/Data/Set/Ncard.lean [hunk 6]
```diff
@@ -218,11 +621,7 @@ theorem pred_ncard_le_ncard_diff_singleton (s : Set α) (a : α) : s.ncard - 1 
 #align set.pred_ncard_le_ncard_diff_singleton Set.pred_ncard_le_ncard_diff_singleton
 
 theorem ncard_exchange (ha : a ∉ s) (hb : b ∈ s) : (insert a (s \ {b})).ncard = s.ncard := by
-  cases' s.finite_or_infinite with h h
-  · haveI := h.to_subtype
-    rw [ncard_insert_of_not_mem, ncard_diff_singleton_add_one hb]
-    simp [mem_diff, not_and, not_not, ha];
-  rw [((h.diff (Set.toFinite {b})).mono (subset_insert _ _)).ncard, h.ncard]
+  rw [ncard_def, encard_exchange ha hb, ←ncard_def]
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:624]:
```suggestion
  congr_arg ENat.toNat <| encard_exchange ha hb
```

**Review by j-loreaux — COMMENTED** [review]
This looks pretty nice thanks! I left a bunch of suggestions, but they are minor. Thanks for refactoring `Set.ncard` to use `Set.encard`.

We may want to rename the file (in a different PR probably) to `Data.Set.Card` after this is merged.

**Comment by apnelson1** [comment]
> This looks pretty nice thanks! I left a bunch of suggestions, but they are minor. Thanks for refactoring `Set.ncard` to use `Set.encard`.
> 
> We may want to rename the file (in a different PR probably) to `Data.Set.Card` after this is merged.

Thanks for the comments. I'm on board with the rename - I'll PR it once this one is through.

**Comment by j-loreaux** [comment]
@apnelson1 just FYI, I think you may have missed a bunch of the comments (GitHub automatically folds long sections of the PR/comment history and you have to click to unfold it).

File: Mathlib/Data/Set/Ncard.lean [hunk 7]
```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
...
+    PartENat.withTopEquiv_symm_zero, PartENat.card_eq_zero_iff_empty, isEmpty_subtype,
+    eq_empty_iff_forall_not_mem]
+
+@[simp] theorem encard_empty : (∅ : Set α).encard = 0 := by
+  rw [encard_eq_zero]
+
+theorem nonempty_of_encard_ne_zero (h : s.encard ≠ 0) : s.Nonempty := by
+  rwa [nonempty_iff_ne_empty, Ne.def, ←encard_eq_zero]
+
+theorem encard_ne_zero : s.encard ≠ 0 ↔ s.Nonempty := by
```
[the remark this answers is not in the document]
Comment (apnelson1) [reply Mathlib/Data/Set/Ncard.lean:100]: `simp` already reduces the LHS to `¬(s = ∅)` so the linter dislikes this.

**Comment by apnelson1** [comment]
> @apnelson1 just FYI, I think you may have missed a bunch of the comments (GitHub automatically folds long sections of the PR/comment history and you have to click to unfold it).

Oops/Thanks! - I think they're all addressed now.

File: Mathlib/Data/Set/Ncard.lean [hunk 8]
```diff
@@ -361,46 +378,42 @@ section Function
...
-  encard_image_of_injOn (hf.injOn s)
+theorem _root_.Function.Injective.encard_image (hf : f.Injective) (s : Set α) :
+    (f '' s).encard = s.encard :=
+  (hf.injOn s).encard_image
 
-theorem encard_le_of_embedding (e : s ↪ t) : s.encard ≤ t.encard := by
-  rw [←encard_univ_coe, ←encard_image_of_injective e.injective univ,
-    ←encard_image_of_injective Subtype.coe_injective]
+theorem _root_.Function.Embedding.enccard_le (e : s ↪ t) : s.encard ≤ t.encard := by
+  rw [←encard_univ_coe, ←(e.injective).encard_image, ←(Subtype.coe_injective).encard_image]
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:389]:
```suggestion
  rw [←encard_univ_coe, ←e.injective.encard_image, ←Subtype.coe_injective.encard_image]
```

File: Mathlib/Data/Set/Ncard.lean [hunk 9]
```diff
@@ -432,25 +445,24 @@ theorem Finite.exists_bijOn_of_encard_eq [Nonempty β] (hs : s.Finite) (h : s.en
   obtain ⟨f, hf, hinj⟩ := hs.exists_injOn_of_encard_le h.le; use f
   convert hinj.bijOn_image
   rw [(hs.image f).eq_of_subset_of_encard_le' (image_subset_iff.mpr hf)
-    (h.symm.trans (encard_image_of_injOn hinj).symm).le]
+    (h.symm.trans (hinj.encard_image).symm).le]
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:445]:
```suggestion
    (h.symm.trans hinj.encard_image.symm).le]
```

**Review by j-loreaux — APPROVED** [review]
Two sets of extra parentheses left, but I'm delegating this to you. Please don't merge before CI finishes.

**Comment by j-loreaux** [comment]
bors d+

**Pull request closed without merging**